cpp/ql/lib/semmle/code/cpp/ir/implementation/aliased_ssa/Operand.qll

```diff
@@ -416,10 +416,9 @@ class ThisArgumentOperand extends ArgumentOperand {
   // in most cases the def location makes more sense, but in some corner cases it
   // does not have a location: in those cases we fall back to the use location
   override Language::Location getLocation() {
-    result = this.getAnyDef().getLocation()
-    or
-    not exists(this.getAnyDef().getLocation()) and
-    result = this.getUse().getLocation()
+    if not this.getAnyDef().getLocation() instanceof Language::UnknownLocation
+    then result = this.getAnyDef().getLocation()
+    else result = this.getUse().getLocation()
```

Comment:
This `not` in the condition is blowing my mind a bit 🤔. I think there are three conditions to think about:
1. What happens when `this.getAnyDef().getLocation() instanceof Language::UnknownLocation` holds.
2. What happens when `this.getAnyDef().getLocation()` doesn't exist.
3. What happens when `this.getAnyDef().getLocation()` exists and `this.getAnyDef().getLocation() instanceof Language::UnknownLocation` doesn't hold.

For 1.: When `this.getAnyDef().getLocation() instanceof Language::UnknownLocation` holds, then we pick the location given by `this.getUse().getLocation()`. That's perfect!
For 2.: When `this.getAnyDef().getLocation()` doesn't exist, then `not this.getAnyDef().getLocation() instanceof Language::UnknownLocation` holds, and then we pick the the location given by `this.getAnyDef().getLocation()`. But this one doesn't exist. That seems like a problem, I think.
For 3.: When `this.getAnyDef().getLocation()` exists and `this.getAnyDef().getLocation() instanceof Language::UnknownLocation` doesn't hold, we pick the location given by `this.getUse().getLocation()`, right? That also seems wrong to me?

Reply:
As far as I can see in 3. when `this.getAnyDef().getLocation() instanceof Language::UnknownLocation` does not hold we correctly choose `this.getAnyDef().getLocation()`.

As for 2., I think that this probably happens to work because there might be no `Instruction` without a `Location`, and at least I could not find any in the ir test folder. But I could not really confirm this from the code itself, so maybe better be ready for it in any case.

Reply:
I do still need to wrap my head around the not existing semantics though :laughing:

Reply:
Oops. You're right wrt. 3. I confused myself! Yes, when `this.getAnyDef().getLocation()` exists and `this.getAnyDef().getLocation() instanceof Language::UnknownLocation` doesn't hold, then `not this.getAnyDef().getLocation() instanceof Language::UnknownLocation` holds, and thus we pick `this.getAnyDef().getLocation()` as expected ✅.

Reply:
I hope the current version is clearer, what do you think?

Reply:
> As for 2., I think that this probably happens to work because there might be no `Instruction` without a `Location`, and at least I could not find any in the ir test folder. But I could not really confirm this from the code itself, so maybe better be ready for it in any case.

In theory, every `Instruction` should have a result for `getLocation()` (because they should have an `UnknownLocation` result if there's no good location to pick). So maybe this is actually good, and we can fix the missing locations in another PR if they show up later. How does that sound to you?